### ISSUE TITLE
train imagenet

### ISSUE THREAD

**Issue by congmonkey** [bob48523/fine-tune-tri-tunnel-net#2]
# -*- coding: utf-8 -*-

# License: BSD
# Author: Sasank Chilamkurthy

from __future__ import print_function, division

import argparse

from models import * 
#import config as cf
from preprocess import *

import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd import Variable
import torch.backends.cudnn as cudnn

import torchvision
from torchvision import datasets, models, transforms

import matplotlib.pyplot as plt
import time
import copy
import os
import setproctitle
import numpy as np

os.environ["CUDA_VISIBLE_DEVICES"] = "2"

parser = argparse.ArgumentParser(description='PyTorch Imagenet Training')
parser.add_argument('--resume', '-r', action='store_true', help='resume from checkpoint')
parser.add_argument('--save')
parser.add_argument('--lr', default=0.1, type=float)
parser.add_argument('--epoch', default=100, type=int)
parser.add_argument('--wd', default=5e-4, type=float)
parser.add_argument('--momentum', default=0.9, type=float)
parser.add_argument('--testOnly', '-t', action='store_true', help='Test mode with the saved model')

args = parser.parse_args()
args.save = args.save or './tiny/tiny_tune1.base'
setproctitle.setproctitle(args.save)

use_cuda = torch.cuda.is_available()
best_acc = 0  # best test accuracy
start_epoch = 1  # start from epoch 0 or last checkpoint epoch
batch_size = 256
epochs = args.epoch
lr = args.lr
weight_decay = args.wd
momentum = args.momentum

if not os.path.exists(args.save):
    os.mkdir(args.save)
######################################################################
# Load Data

data_transforms = {
    'train': transforms.Compose([
        transforms.RandomSizedCrop(64),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize([0.5,0.5,0.5], [0.5,0.5,0.5])
    ]),
    'val': transforms.Compose([
        transforms.Scale(72),
        transforms.CenterCrop(64),
        transforms.ToTensor(),
        transforms.Normalize([0.5,0.5,0.5], [0.5,0.5,0.5])
    ]),
}
data_dir = '/ssd/RookieProject/TinyImageNet'
#data_dir = '/ssd/chenjiaxu/data/hymenoptera_data'
trainset = datasets.ImageFolder(os.path.join(data_dir, 'train'), data_transforms['train'])
valset = datasets.ImageFolder(os.path.join(data_dir, 'val'), data_transforms['val'])       
trainloader = torch.utils.data.DataLoader(trainset, batch_size=batch_size,
                                               shuffle=True, num_workers=4)
                
valloader = torch.utils.data.DataLoader(valset, batch_size=batch_size,
                                               shuffle=False, num_workers=4)

#data_dir = '/ssd/RookieProject/CIFAR100/'
#trainset = TxtFolder(os.path.join(data_dir), 'list_train.txt', data_transforms['train'])
#
#trainloader =torch.utils.data.DataLoader(trainset, batch_size=cf.batch_size,
#                                               shuffle=True, num_workers=4)
#
#valset = TxtFolder(os.path.join(data_dir), 'list_val.txt', data_transforms['val'])
#
#valloader =torch.utils.data.DataLoader(trainset, batch_size=cf.batch_size,
#                                               shuffle=False, num_workers=4)
              
#dset_sizes = {x: len(dsets[x]) for x in ['train', 'val']}
dset_classes = trainset.classes
num_classes = len(dset_classes)
use_gpu = torch.cuda.is_available()

class AverageMeter(object):
    """Computes and stores the average and current value"""
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count
######################################################################
# Training the model
# Test only option
if (args.testOnly):
    print('\n[Test Phase] : Model setup')
    #assert os.path.isdir('checkpoint'), 'Error: No checkpoint directory found!'
    checkpoint = torch.load(os.path.join(args.save, 'imagenet.t7'))
    net = checkpoint['net']
    epoch = checkpoint['epoch']
    print (epoch)
    if use_cuda:
        net.cuda()
       # net = torch.nn.DataParallel(net, device_ids=range(torch.cuda.device_count()))
       # cudnn.benchmark = True

    net.eval()
    test_loss = 0
    correct = 0
    total = 0

    for batch_idx, (inputs, targets) in enumerate(valloader):
        if use_cuda:
            inputs, targets = inputs.cuda(), targets.cuda()
        inputs, targets = Variable(inputs, volatile=True), Variable(targets)
        outputs = net(inputs)

        _, predicted = torch.max(outputs.data, 1)
        total += targets.size(0)
        correct += predicted.eq(targets.data).cpu().sum()

    # Save checkpoint when best model
    acc = 100.*correct/total
    print("| Test Result\tAcc@1: %.2f%%" %(acc))

    sys.exit(0)
    
def train(net, criterion, epoch, trainF):
    print('\nEpoch: %d' % epoch)
    net.train()
    train_loss = 0
    top1 = AverageMeter()
    top5 = AverageMeter()
    losses = AverageMeter()
#    correct = 0
    total = 0
    nProcessed = 0
    
    nTrain = len(trainloader.dataset)

    optimizer = optim.SGD(net.parameters(), lr=lr, momentum=momentum, weight_decay=weight_decay)
    
    for batch_idx, (inputs, targets) in enumerate(trainloader):
        if use_cuda:
            inputs, targets = inputs.cuda(), targets.cuda()
        optimizer.zero_grad()
        target = targets
        inputs, targets = Variable(inputs), Variable(targets)
        outputs = net(inputs)
        loss = criterion(outputs, targets)
        #make_graph.save('./t.dot', loss.creator); assert(False)
        loss.backward()
        optimizer.step()
        
        nProcessed += len(inputs)
        train_loss += loss.data[0]
        _, predicted = torch.max(outputs.data, 1)
        total += targets.size(0)
#        correct += predicted.eq(targets.data).cpu().sum()
        prec1, prec5 = accuracy(outputs.data, target, topk=(1, 5))
        top1.update(prec1[0], inputs.size(0))
        top5.update(prec5[0], inputs.size(0))
        losses.update(loss.data[0], inputs.size(0))
        incorrect = predicted.ne(targets.data).cpu().sum()
        err = 100.*incorrect/len(inputs)
        partialEpoch = epoch + batch_idx / len(trainloader) - 1
        
        print('Train Epoch: {:.2f} [{}/{} ({:.0f}%)]\tLoss: {:.6f}\tError: {:.6f}'.format(
            partialEpoch, nProcessed, nTrain, 100. * batch_idx / len(trainloader),
            loss.data[0], err))
        print(' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}'
              .format(top1=top1, top5=top5))
        
    trainF.write('{},{},{},{}\n'.format(epoch, losses.avg, top1.avg, top5.avg))
    trainF.flush()
        
def test(net, epoch, testF):
    global best_acc
    net.eval()
    top1 = AverageMeter()
    top5 = AverageMeter()
    test_loss = 0
    incorrect = 0
    correct = 0
    total = 0 
    
    for batch_idx, (inputs, targets) in enumerate(valloader):
        if use_cuda:
            inputs, targets = inputs.cuda(), targets.cuda()
        target = targets
        inputs, targets = Variable(inputs, volatile=True), Variable(targets)
        outputs = net(inputs)
        loss = criterion(outputs, targets)

        test_loss += loss.data[0]
        _, predicted = torch.max(outputs.data, 1)
        total += targets.size(0)
        correct += predicted.eq(targets.data).cpu().sum()
        incorrect += predicted.ne(targets.data).cpu().sum()
        prec1, prec5 = accuracy(outputs.data, target, topk=(1, 5))
        top1.update(prec1[0], inputs.size(0))
        top5.update(prec5[0], inputs.size(0))

    nTotal = len(valloader.dataset)
    err  = 100.*incorrect/nTotal
    test_loss /= len(valloader)
    
    print('\nTest set: Average loss: {:.4f}, Error: {}/{} ({:.0f}%)\n'.format(
        test_loss, incorrect, nTotal, err))
    print(' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}'
          .format(top1=top1, top5=top5))
    testF.write('{},{},{},{}\n'. format(epoch, test_loss, top1.avg, top5.avg))
    testF.flush()   

    # Save checkpoint.
    acc = 100.*correct/total
    
    state = {
        'net': net.module if use_cuda else net,
        #'net': net,
        'acc': acc,
        'epoch': epoch,
    }
    torch.save(state, os.path.join(args.save, 'imagenet_latest.t7'))
    if acc > best_acc:
        torch.save(state, os.path.join(args.save, 'imagenet.t7'))
        best_acc = acc 
            

######################################################################
# Visualizing the model predictions
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#
# Generic function to display predictions for a few images
#

def accuracy(output, target, topk=(1,)):
    """Computes the precision@k for the specified values of k"""
    maxk = max(topk)
    batch_size = target.size(0)

    _, pred = output.topk(maxk, 1, True, True)
    pred = pred.t()
    correct = pred.eq(target.view(1, -1).expand_as(pred))

    res = []
    for k in topk:
        correct_k = correct[:k].view(-1).float().sum(0)
        res.append(correct_k.mul_(100.0 / batch_size))
    return res

######################################################################
# Finetuning the convnet
# ----------------------
# Model
if args.resume:
    # Load checkpoint.
    print('==> Resuming from checkpoint..')
    #assert os.path.isdir('checkpoint'), 'Error: no checkpoint directory found!'
    checkpoint = torch.load(os.path.join(args.save, 'imagenet_latest.t7'))
    model_ft = checkpoint['net']
    best_acc = checkpoint['acc']
    start_epoch = checkpoint['epoch']+1
else:
    print('==> Building new model..')
    #model_ft = Resnet_drop(PreActBlock, [2,2,2,2], num_classes)
    model_ft = Resnet_tune(PreActBlock, [2,2,2,2], num_classes)
    print (model_ft)
    #net = ResNet34()
    print ('Number of parames:{}'.format(
        sum([p.data.nelement() for p in model_ft.parameters()])))
    
if use_gpu:
    model_ft = model_ft.cuda()
    model_ft = torch.nn.DataParallel(model_ft, device_ids=[0])
    cudnn.benchmark = True

criterion = nn.CrossEntropyLoss()

######################################################################
# Train and evaluate
# ^^^^^^^^^^^^^^^^^^

if args.resume:
    trainF = open(os.path.join(args.save, 'train.csv'), 'a')
    testF = open(os.path.join(args.save, 'test.csv'), 'a')
else:
    trainF = open(os.path.join(args.save, 'train.csv'), 'w')
    testF = open(os.path.join(args.save, 'test.csv'), 'w')  
    
elapsed_time = 0
#epochs = 300
for epoch in range(start_epoch, start_epoch+epochs):
    start_time = time.time()
    train(model_ft, criterion, epoch, trainF)
    test(model_ft, epoch, testF)
    epoch_time = time.time() - start_time
    elapsed_time += epoch_time
    #print('| Elapsed time : %d:%02d:%02d'  %(cf.get_hms(elapsed_time)))
    
trainF.close()
testF.close()
